mm-large/mm-large.c

```diff
@@ -0,0 +1,344 @@
+
+      if (thread_id == 0) {
+        // swap buffer
+        a_idx = a_idx ? 0 : 1;
+
+        if (s < n_stripes-1) {
+          // determine next DMA XFER
+          const unsigned ext_addr = (unsigned)a + (s+1)*stripe_size_b;
+
+          // set up DMA XFER
+          a_dma[a_idx] = hero_dma_memcpy_async((void *)a_ptrs[a_idx], (void *)ext_addr, stripe_size_b);
+        }
+
+        // wait for previous DMA XFER
+        hero_dma_wait(a_dma[!a_idx]);
+      }
+      else if ( (thread_id == 2) && (s > 0) ) {
+        // swap buffer
+        c_idx = c_idx ? 0 : 1;
+
+        // determine next DMA XFER
+        const unsigned ext_addr = (unsigned)c + (s-1)*stripe_size_b;
+
+        // set up DMA XFER
+        c_dma[!c_idx] = hero_dma_memcpy_async((void *)ext_addr, (void *)c_ptrs[!c_idx], stripe_size_b);
+
+        // wait for previous DMA XFER
+        if (s > 1)
+          hero_dma_wait(c_dma[c_idx]);
+      }
+
+      // vertical b stripes
+      for (int t=0; t<n_stripes; t++) {
+
+        if ( (thread_id == 1) ) {
+          // swap buffer
+          b_idx = b_idx ? 0 : 1;
+
+          if (t < n_stripes-1) {
+            // determine next DMA XFER
+            const unsigned ext_addr = (unsigned)b + (t+1)*stripe_size_b;
+
+            // set up DMA XFER
+            b_dma[b_idx] = hero_dma_memcpy_async((void *)b_ptrs[b_idx], (void *)ext_addr, stripe_size_b);
+          }
+          else if (s < n_stripes-1) {
+            // determine next DMA XFER
+            const unsigned ext_addr = (unsigned)b;
+
+            // set up DMA XFER
+            b_dma[b_idx] = hero_dma_memcpy_async((void *)b_ptrs[b_idx], (void *)ext_addr, stripe_size_b);
+          }
+
+          // wait for previous DMA XFER
+          hero_dma_wait(b_dma[!b_idx]);
+        }
+
+        #pragma omp barrier
+
+        #pragma omp for collapse(2)
+
+        // horizontal a and c rows
+        for (int i=0; i<stripe_height_local; i++) {
+
+          // vertical b columns
+          for (int j=0; j<stripe_height_local; j++) {
+
+            int sum = 0;
+            for (int k=0; k<width_local; k++) {
+              sum = sum + a_ptrs[!a_idx][i*width_local+k] * b_ptrs[!b_idx][j*width_local+k];
+            } // k < width_local
+            c_ptrs[c_idx][i*width_local+t*stripe_height_local+j] = sum;
+          } // j < stripe_height_local
+        } // i < stripe_height_local
+      } // t < n_stripes
+
+    } // n_stripes
+
+    // copy out last c stripe
+    if (thread_id == 2)
+      hero_dma_memcpy((void *)((unsigned)c+(n_stripes-1)*stripe_size_b), (void *)c_ptrs[c_idx], stripe_size_b);
+
+  } // parallel
+
+  hero_l1free(a_ptrs[0]);
+  hero_l1free(a_ptrs[1]);
+  hero_l1free(b_ptrs[0]);
+  hero_l1free(b_ptrs[1]);
+  hero_l1free(c_ptrs[0]);
+  hero_l1free(c_ptrs[1]);
+
+  return 0;
+}
+
+#pragma omp end declare target
+
+int main(int argc, char *argv[])
+{
+  printf("HERO matrix multiplication started.\n");
+
+  // Global variables
+  uint32_t *a;
```

Comment:
I suggest to declare and define these in one line as
```c
uint32_t* const a = (uint32_t*)malloc(width * height * sizeof(uint32_t));
```
Reasons: Increased const-correctness and zero declaration-definition distance.